File: freight-triage-polyglot-ledger-recovery/environment/reconcile/internal/statx/absdev_milli.go
package statx

// AbsdevMilli is the absdev_milli kernel.
func AbsdevMilli(series []int64) int64 {
	if len(series) == 0 {
		return 0
	}
	count := int64(len(series))
	total := int64(0)
	for _, value := range series {
		total += value
	}
	mean := FloorDiv(total*1000, count)
	accumulator := int64(0)
	for _, value := range series {
		delta := value*1000 - mean
		if delta < 0 {
			delta = -delta
		}
		accumulator += delta
	}
	return FloorDiv(accumulator, count)
}

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/stats/AbsdevMilli.java
package com.freight.stats;

/** absdev_milli kernel. */
public final class AbsdevMilli implements StatKernel {

    @Override
    public String name() {
        return "absdev_milli";
    }

    @Override
    public long apply(long[] series) {
        if (series.length == 0) {
            return 0L;
        }
        long count = series.length;
        long total = 0L;
        for (int i = 0; i < series.length; i++) {
            total += series[i];
        }
        long mean = StatSupport.floorDiv(total * 1000L, count);
        long accumulator = 0L;
        for (int i = 0; i < series.length; i++) {
            long delta = series[i] * 1000L - mean;
            accumulator += delta < 0L ? -delta : delta;
        }
        return StatSupport.floorDiv(accumulator, count);
    }
}
